lib/datadog/core/telemetry/collector.rb

```diff
@@ -0,0 +1,241 @@
+            !Bundler.bundle_path.nil?
+          rescue Bundler::GemfileNotFound
+            false
+          end
+        end
+
+        def env
+          Datadog.configuration.env
+        end
+
+        def service_name
+          Datadog.configuration.service
+        end
+
+        def service_version
+          Datadog.configuration.version
+        end
+
+        def tracer_version
+          Core::Environment::Identity.tracer_version
+        end
+
+        def products
+          profiler_obj = profiler
+          appsec_obj = appsec
+          profiler_obj || appsec_obj ? Telemetry::V1::Product.new(profiler: profiler_obj, appsec: appsec_obj) : nil
+        end
+
+        def profiler
+          version = profiler_version
+          Telemetry::V1::Profiler.new(version: version) if version
+        end
+
+        def profiler_version
+          tracer_version if Datadog.configuration.respond_to?(:profiling) && Datadog.configuration.profiling.enabled
+        end
+
+        def appsec
+          version = appsec_version
+          Telemetry::V1::AppSec.new(version: version) if version
+        end
+
+        def appsec_version
+          tracer_version if Datadog.configuration.respond_to?(:appsec) && Datadog.configuration.appsec.enabled
+        end
+
+        def configuration_variables
+          configurations = []
+          environment_collector = Core::Diagnostics::EnvironmentCollector.new
+          [
+            Telemetry::V1::Configuration.new(
+              name: 'DD_AGENT_HOST', value: Datadog.configuration.agent.host || ENV.fetch('DD_AGENT_HOST', '127.0.0.1')
+            ),
+            Telemetry::V1::Configuration.new(name: 'DD_AGENT_TRANSPORT', value: agent_transport),
+            Telemetry::V1::Configuration.new(name: 'DD_TRACE_AGENT_URL', value: environment_collector.agent_url),
+            Telemetry::V1::Configuration.new(
+              name: 'DD_TRACE_SAMPLE_RATE',
+              value: environment_collector.sample_rate || Datadog.configuration.tracing.sampling.default_rate
+            )
+          ].each do |configuration|
+            configurations << configuration unless configuration.value.nil?
+          end
+          configurations
+        end
+
+        def agent_transport
+          if !!ENV.fetch('DD_APM_RECEIVER_SOCKET', nil)
+            'UDS'
+          else
+            'TCP'
+          end
+        end
+
+        def flatten_configuration(hash, configuration_array)
+          flattened_hash = flatten_hash(hash)
+          flattened_hash.each do |k, v|
+            configuration_array << Telemetry::V1::Configuration.new(name: k.to_s, value: v)
+          end
+        end
+
+        def flatten_hash(hash)
+          hash.to_h.each_with_object({}) do |(k, v), h|
+            if empty?(v) || (v.is_a? Array) || (v.is_a? String) || (v.is_a? Integer) || (v.is_a? Float)
+              next
+            elsif v.respond_to?(:to_h) && !v.to_h.empty?
+              flatten_hash(v.to_h).map do |h_k, h_v|
```

Comment:
`#to_h` can be expensive, and we are calling it twice for non-empty hashes, we should cache this value:
```suggestion
            elsif v.respond_to?(:to_h) && !(inner_hash = v.to_h).empty?
              flatten_hash(inner_hash).map do |h_k, h_v|
```

If our friend Rubocop yells at us for this, just follow its lead 🤷